### ISSUE TITLE
Imageset dataset partition

### ISSUE THREAD

**Issue by cveres** [NervanaSystems/neon#72]
I have trained a network with the Asirra/Petfinder dataset with cats and dogs.
The images are arranged in two labeled folders, since the documentation says: " It can also take advantage of directory subfolders to identify target labels."

The documentation is not so clear about the test set, so I created test folders to mimic the training data. Here is my layout:

~data/Images/train/Cat
-------------------------/Dog
~data/Images/test/Cat
------------------------/Dog

However running neon against my yaml file ends with this error:

2015-08-13 23:39:08,885 WARNING:fit_predict_err - Unable to generate 'test' metrics, no equivalent dataset partition

So, how should the test data be partitioned?

Here are some more lines of the output, as well as the appropriate section from the Yaml file.
You will see in the Yaml file I tried manipulating these settings, but it does not have any effect on the error:
inference_sets: ['validate','train'],
 # predictions: ['train', 'test'],

2015-08-13 23:38:05,539 INFO:mlp - epoch: 18, training error: 0.65325
2015-08-13 23:38:36,876 INFO:mlp - epoch: 19, training error: 0.65011
2015-08-13 23:39:08,099 INFO:mlp - epoch: 20, training error: 0.64382
2015-08-13 23:39:08,139 WARNING:persist - serializing object to: /Users/csabaveres/data/Images/ndsb-model.prm
2015-08-13 23:39:08,885 WARNING:fit_predict_err - Unable to generate 'test' metrics, no equivalent dataset partition
2015-08-13 23:39:21,870 INFO:fit_predict_err - train set MisclassPercentage_TOP_1 33.29327

YAML:

!obj:experiments.FitPredictErrorExperiment {
  inference_sets: ['validate','train'],
 # predictions: ['train', 'test'],

  dataset: &ds !obj:datasets.Imageset {
    repo_path: '~/data',
    imageset: 'Images',
    square_crop: True,
    image_dir: '~/data/Images/train', 
    save_dir: '~/data/Images/batches',
    output_image_size: &ois 128,
    cropped_image_size: &cis 64,
    macro_size: 1024,
    num_channels: 1,
  },

  metrics: {
    train: [
      !obj:metrics.MisclassPercentage {},
    ],
    test: [
      !obj:metrics.MisclassPercentage {},
    ],
  },

thanks!

**Comment by scttl** [comment]
Hi,

Unfortunately neon's `BatchWriter` class (which is what `Imageset` based datasets use for a lot of the initial input file prep) doesn't currently operate in the manner you expect.  Today the base `BatchWriter` class only supports a single top-level input directory, with subdirectories for target label classes.  This directory of images gets read and randomly slotted into 80% train and 20% validation sets by default (though you can change this mix via the `validation_pct` parameter for `Imageset`).

So if your test set happens to be an arbitrary selected holdout set, the fastest way for you to get going would be to lump all your original train/test split data together under a single root directory and let `BatchWriter` do the splitting for you -- you'd also need to replace the `test` based metrics with `validation` metrics in your YAML file.

If you specifically want to keep your already defined test set images as a test set, you're going to have to write your own subclass of `BatchWriter` and make a change or two to `Imageset` to ensure it loads your derived `BatchWriter` class.
